pc-parse/pcpatr/grammar.c

```diff
@@ -1973,6 +2060,180 @@ listp->pNext      = pData->pGrammar->pRuleTable;
+ *    pConstraints_in
+ *    pData   - grammar data
+ * DESCRIPTION
+ *    Install two rules, one with nterm and one without.
+ * RETURN VALUE
+ *    void
+ */
+static void expand_optional_non_terminal(nterm, id, lhs, psr, dag,
+	pPriorityUnions_in, pConstraints_in, pData)
+PATRNonterminal * nterm;
+char *			id;
+char *			lhs;
+SimplePSR *         psr;
+PATRFeature *		dag;
+PATRPriorityUnion *	pPriorityUnions_in;
+PATRConstraint *	pConstraints_in;
+GrammarData *		pData;
+{
+SimplePSR *psr2;
+PATRFeature * dag2;
+PATRNonterminal * nterm2;
+/* Make a rule without nterm. */
+psr2 = copy_psr(psr, pData);
+if (psr2->pHead->pszName == nterm->pszName) {
+	/* Remove the non-terminal at the beginning. */
+	psr2->pHead = psr2->pHead->pNext;
+	if (psr->pHead == psr->pTail) {
+		/* Update end of list pointer. */
+		psr2->pTail = NULL;
+	}
+} else {
+	/* Remove an embedded non-terminal. */
+	for( nterm2 = psr2->pHead ; nterm2 ; nterm2 = nterm2->pNext ) {
+	if (nterm2->pNext->pszName == nterm->pszName) {
+		nterm2->pNext = nterm2->pNext->pNext;
+		break;
+	}
+	}
+}
+dag2 = skip_optional_attr(dag, nterm->pszName, TRUE, pData->pPATR);
+install_rule(id, lhs, psr2, dag2, pPriorityUnions_in, pConstraints_in, pData);
+/* Make a rule with nterm obligatory. */
+psr2 = copy_psr(psr, pData);
+for( nterm2 = psr2->pHead ; nterm2 ; nterm2 = nterm2->pNext ) {
+if (nterm2->pszName == nterm->pszName) {
+	nterm2->bOptional = FALSE;
+	break;
+}
+}
+install_rule(id, lhs, psr2, dag, pPriorityUnions_in, pConstraints_in, pData);
+}
+
+/*****************************************************************************
+ * NAME
+ *    skip_optional_attr
+ * ARGUMENTS
+ *    pDag - feature structure
+ *    attr - attribute
+ *    exclude - whether to exclude or include optional constraints
+ *    pPATR - PATR data
+ * DESCRIPTION
+ *    Reunify pDag from constraints skipping constraints that mention attr.
+ * RETURN VALUE
+ *    PATRFeature *
+ */
+PATRFeature *skip_optional_attr(pDag, attr, exclude, pPATR)
+PATRFeature *	pDag;
+char * attr;
+int exclude;
+PATRData * pPATR;
+{
+/*
+ * We can retroactively skip constraints that mention attr
+ * because the unifier preserved the sources of each unification.
+ * This is too complicated for the parser to do since its DAG
+ * is created by unifying active and passive edges on top
+ * of the rule constraints.
+ */
+PATRFeature *pFirstFeat;
+PATRFeature *pSecondFeat;
+PATRFeature *pDag2;
+/* See if pDag has attr. */
+int has_attr = FALSE;
+PATRComplexFeature *flist;
+if (pDag->pFirstFeat || pDag->pSecondFeat) {
+	/* Unify pDag again. */
+	pFirstFeat = skip_optional_attr(pDag->pFirstFeat, attr, exclude, pPATR);
+	pSecondFeat = skip_optional_attr(pDag->pSecondFeat, attr, exclude, pPATR);
+	if (!pFirstFeat) {
+		return pSecondFeat;
+	}
+	if (!pSecondFeat) {
+		return pFirstFeat;
+	}
+	pDag2 = unifyPATRFeatures(pFirstFeat, pSecondFeat, TRUE, pPATR);
+	return pDag2;
+}
+for (flist = pDag->u.pComplex ; flist ; flist = flist->pNext) {
```

Comment:
should we check for pDag->eType == PATR_COMPLEX before launching into this for loop?

Reply:
I think that pDag->eType is always PATR_COMPLEX at this point since pDag is not the result of a unification, it is the result of reading a constraint in from the grammar (if it were the result of a unification, then pDag->pFirstFeat or pDag->pSecondFeat would be non-NULL).

Reply:
I will add if (pDag->eType == PATR_COMPLEX) ... just to be sure.